user/config/config_gotls.go

```diff
@@ -340,7 +340,7 @@ func (gc *GoTLSConfig) findRetOffsetsPie(lfunc string) ([]int, error) {
 			continue
 		}
 		data := make([]byte, funcLen)
-		_, err = prog.ReadAt(data, int64(address))
+		_, err = prog.ReadAt(data, int64(address-prog.Vaddr))
```

Comment:
In [this code](https://github.com/golang/go/blob/bdd27c4debfb51fe42df0c0532c1c747777b7a32/src/cmd/internal/objfile/elf.go#L175):
```
addr = s(type: *elf.Symbol).Value 
prog.ReadAt(data, addr- prog.Vaddr )
```
However, in our code:
```
address, err = gc.findPieSymbolAddr(lfunc)
    (address = f(type: *gosym.Function).Value - prog.Vaddr + prog.Off)
_, err = prog.ReadAt(data, int64(address-prog.Vaddr))
```
We know that the virtual address from the symbol is the same as the virtual address from gosym.Function. Therefore, I believe there is an error, we should use `f.Value` instead.
I check this in my armLinux, use `f.Value` is correct, and `address` is wrong.

However, I don't understand why it runs properly in your test, LOL.

Reply:
in `address := f.Value - prog.Vaddr + prog.Off`,  `prog.Vaddr` and `prog.Off` are the same value, so the calculation here is meaningless (removed).

So, taking your advice into consideration, I just returned `f.Value`.

### ARM64 ubuntu 22.04
`prog.Vaddr` ==`prog.Off` == 0

```shell
FUnc s.Name:crypto/tls.(*Conn).writeRecordLocked, s.Value:9FF0B0, prog.Vaddr:0, prog.Off:0
FUnc s.Name:crypto/tls.(*Config).writeKeyLog, s.Value:9FA460, prog.Vaddr:0, prog.Off:0
FUnc s.Name:crypto/tls.(*Conn).Read, s.Value:A014A0, prog.Vaddr:0, prog.Off:0
RetOffsetPie address:A014A0, prog.Vaddr:0,  prog.info:&{{PT_LOAD PF_X+PF_R 0 0 0 34623968 34623968 65536} 0x4000088930 0x4000088930}
```


### AMD64 ubuntu 22.04
`prog.Vaddr` ==`prog.Off` == 14C4000

```shell
FUnc s.Name:crypto/tls.(*Conn).writeRecordLocked, s.Value:19E64C0, prog.Vaddr:14C4000, prog.Off:14C4000
FUnc s.Name:crypto/tls.(*Config).writeKeyLog, s.Value:19E0380, prog.Vaddr:14C4000, prog.Off:14C4000
FUnc s.Name:crypto/tls.(*Conn).Read, s.Value:19E8EA0, prog.Vaddr:14C4000, prog.Off:14C4000
func:crypto/tls.(*Conn).Read, address:19e8ea0
RetOffsetPie address:19E8EA0, prog.Vaddr:14C4000,  prog.info:&{{PT_LOAD PF_X+PF_R 14C4000 14C4000 14C4000 0x1BE322D 0x1BE322D 4096} 0xc00007aa80 0xc00007aa80}
```

Reply:
and now, Its calculation logic is the same as [this code](https://github.com/golang/go/blob/bdd27c4debfb51fe42df0c0532c1c747777b7a32/src/cmd/internal/objfile/elf.go#L175).